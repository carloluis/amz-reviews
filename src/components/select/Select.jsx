import React from 'react';
import PropTypes from 'prop-types';
import { classes } from '../../utils';
import styles from './Select.scss';

class Select extends React.PureComponent {
    constructor(props) {
        super(props);

        this.handleSelect = this.handleSelect.bind(this);
        this.toggle = this.toggle.bind(this);
        this.state = {
            open: false
        };
    }
    toggle() {
        if (!this.props.disabled) {
            this.setState(({ open }) => ({ open: !open }));
        }
    }
    handleSelect(event) {
        const value = event.target.dataset.value;
        this.props.onSelect(value);
        this.toggle();
    }
    render() {
        const { selected, title, values, disabled } = this.props;
        const { open } = this.state;

        return (
            <div className={classes(styles.container, disabled && styles.disabled)}>
                <div className={styles.input} onClick={this.toggle}>
                    {selected || title} <i className={styles.triangle}>{open ? '▲' : '▼'}</i>
                </div>
                {open && <div className={styles.backdrop} onClick={this.toggle} />}
                {open && (
                    <div className={styles.options}>
                        {values.map(value => (
                            <div key={value} className={styles.option} onClick={this.handleSelect} data-value={value}>
                                {value}
                            </div>
                        ))}
                    </div>
                )}
            </div>
        );
    }
}

Select.propTypes = {
    disabled: PropTypes.bool,
    onSelect: PropTypes.func.isRequired,
    selected: PropTypes.string,
    title: PropTypes.string.isRequired,
    values: PropTypes.arrayOf(PropTypes.string).isRequired
};

Select.defaultProps = {
    disabled: false,
    selected: ''
};

export default Select;
