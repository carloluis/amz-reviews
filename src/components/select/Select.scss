@import '../../styles/palette.scss';

.container {
    position: relative;
    display: inline-block;
    width: 180px;
    cursor: pointer;
    color: $blue;

    .input {
        position: relative;
        border: 1px solid $blue;
        border-radius: 16px;
        height: 32px;
        padding: 4px 32px 4px 24px;
        font-size: 14px;
        line-height: 22px;

        .triangle {
            position: absolute;
            right: 8px;
        }
    }

    .options {
        position: absolute;
        z-index: 2;
        width: 100%;
        background: $white;
        padding: 4px 0;
        border: 1px solid $blue;
        border-radius: 16px;
        overflow: hidden;
        opacity: 0.9;

        & > div {
            padding: 4px 24px;
            height: 24px;

            &:hover {
                background: $light-grey;
            }
        }
    }

    .backdrop {
        position: fixed;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1;
        cursor: auto;
    }

    &.disabled {
        color: $grey;

        .input {
            border-color: $grey;
        }
    }
}
