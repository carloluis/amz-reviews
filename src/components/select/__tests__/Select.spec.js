import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Select from '../Select';

const PROPS = {
    disabled: false,
    onSelect: jest.fn(),
    selected: '',
    title: 'numbers',
    values: ['one', 'two', 'three', 'four']
};

describe('<Select/>', () => {
    it('snapshot', () => {
        const result = renderer.create(<Select {...PROPS} />).toJSON();
        expect(result).toMatchSnapshot();
    });

    it('snapshot - open', () => {
        const result = renderer.create(<Select {...PROPS} />);
        result.getInstance().setState({ open: true });
        expect(result).toMatchSnapshot();
    });
});

describe('Select', () => {
    let wrapper;
    beforeEach(() => {
        PROPS.onSelect.mockClear();
        wrapper = shallow(<Select {...PROPS} />);

        expect(wrapper.state('open')).toBeFalsy();
    });

    it('should render the options when click on input', () => {
        wrapper.find('.input').simulate('click');
        expect(wrapper.state('open')).toBeTruthy();
    });

    it('should toggle the options by clicking on input', () => {
        for (let time = 0; time < 5; time++) {
            wrapper.find('.input').simulate('click');
            expect(wrapper.state('open'))[time % 2 === 0 ? 'toBeTruthy' : 'toBeFalsy']();
        }
    });

    it('should nnot toggle the options when disabled flag is on', () => {
        wrapper.setProps({ disabled: true });

        for (let time = 0; time < 5; time++) {
            wrapper.find('.input').simulate('click');
            expect(wrapper.state('open')).toBeFalsy();
        }
    });

    it('should call onSelect when click on .option item', () => {
        wrapper.setState({ open: true });

        const event = { target: { dataset: { value: 'one' } } };
        const option = wrapper.find('.option').first();
        option.simulate('click', event);

        expect(PROPS.onSelect).toHaveBeenCalledWith('one');
    });
});
