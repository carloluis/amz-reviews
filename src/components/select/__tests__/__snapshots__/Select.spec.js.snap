// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> snapshot - open 1`] = `
<div
  className="container"
>
  <div
    className="input"
    onClick={[Function]}
  >
    numbers
     
    <i
      className="triangle"
    >
      ▲
    </i>
  </div>
  <div
    className="backdrop"
    onClick={[Function]}
  />
  <div
    className="options"
  >
    <div
      className="option"
      data-value="one"
      onClick={[Function]}
    >
      one
    </div>
    <div
      className="option"
      data-value="two"
      onClick={[Function]}
    >
      two
    </div>
    <div
      className="option"
      data-value="three"
      onClick={[Function]}
    >
      three
    </div>
    <div
      className="option"
      data-value="four"
      onClick={[Function]}
    >
      four
    </div>
  </div>
</div>
`;

exports[`<Select/> snapshot 1`] = `
<div
  className="container"
>
  <div
    className="input"
    onClick={[Function]}
  >
    numbers
     
    <i
      className="triangle"
    >
      ▼
    </i>
  </div>
</div>
`;
